From: Dermot Bradley <dermot_bradley@yahoo.com>
Date: Tue, 6 Sep 2022 23:41:36 +0100
Subject: [PATCH] prevent harmless LVM warnings from grub-probe

Whenever grub-probe is run (by grub-mkconfig) and the root
filesystem is on LVM then multiple warnings may be displayed:

  File descriptor 7 (socket:[24298]) leaked on vgs invocation.
    Parent PID 12489: /usr/sbin/grub-probe

These warnings are harmless but annoying and may make users think
there is an actual problem with their OS. If the environment
variable LVM_SUPPRESS_FD_WARNINGS is set then grub-probe will not
display these. However in the described situation grub-probe is not
being called directly, but rather via grub-mkconfig and so this
patch simply prefixes grub-probe's call to 'vgs' with setting this
environment variable.

I see numerous discussions of this "bug" across multiple distros
(e.g. Debian from 2008, Ubuntu from 2015) over the years and indeed
found two Grub bugs (39696 & 45279) regarding it (from 2013 & 2015)
but no attempt appears to have been made to resolve this upstream.

---

diff -aur a/grub-core/osdep/unix/getroot.c b/grub-core/osdep/unix/getroot.c
--- a/grub-core/osdep/unix/getroot.c
+++ b/grub-core/osdep/unix/getroot.c
@@ -614,7 +614,7 @@
   /* by default PV name is left aligned in 10 character field, meaning that
      we do not know where name ends. Using dummy --separator disables
      alignment. We have a single field, so separator itself is not output */
-  argv[0] = "vgs";
+  argv[0] = "LVM_SUPPRESS_FD_WARNINGS=1 vgs";
   argv[1] = "--options";
   if (vgid)
     argv[2] = "vg_uuid,pv_name";
